modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java

```diff
@@ -830,7 +830,9 @@ void run(BytesReference expected, BytesReference updated, ActionListener<Optiona
 
                 .<Void>newForked(l -> ensureOtherUploadsComplete(uploadId, uploadIndex, currentUploads, l))
 
-                // Step 4: Read the current register value.
+                // Step 4: Read the current register value. Calling getRegister is safe here because all earlier uploads are complete at
+                // this point, our upload is not completing yet, and later uploads can only be completing if they have already aborted ours,
+                // so either this read is linearizable or its result does not matter.
```

Comment:
I believe this. Would it be useful to be explicit about write failures? I wrote this down for myself when I was reading the surrounding code, which I haven't spent much time in:

getRegister only has read-after-write semantics, which means it can return a stale value if another write is in flight, or if the last write failed.
In either of these cases the stale value does not break the correctness of this compare and exchange operation:
* If another write is in flight, then the writer must have first aborted this upload, and so this operation will fail in step 5.
* If the previous write failed, then that writer will fail its compare and exchange operation and nothing will depend on it, so it is safe to overwrite here.

Reply:
> if the last write failed

I am not sure how this could leave a stale value for this writer to read? Seems a serious bug if a failed write still writes to the register? Maybe I am missing something.

Reply:
++ I'm not sure I understand this either. If the previous write failed then it did not modify the contents of the register blob, so we're not really overwriting it.

Reply:
in general there are various things that have to happen after the write is applied before the request completes successfully to the client.

Reply:
I see, ok, and this also applies to e.g. a network outage blocking the response from a write. I opened https://github.com/elastic/elasticsearch/pull/132173 to document this.

> until there is a successful write

This is semantically pretty tricky (and may not even apply to the network outage case).